src/lib/support/Pool.h

```diff
@@ -23,13 +23,20 @@
 
 #pragma once
 
+#include <lib/support/CodeUtils.h>
+
 #include <array>
 #include <assert.h>
 #include <atomic>
 #include <limits>
 #include <new>
 #include <stddef.h>
 
+#if CHIP_SYSTEM_CONFIG_POOL_USE_HEAP
+#include <mutex>
```

Comment:
Might be out of scope for this PR, but can we consider using SystemMutex here? Is there a replacement for set? Basically, std:: stuff makes me a bit nervous because we want larger MPUs to have the option to use HEAP without worrying that some std:: thing will start grabbing up its memory and fragmenting it all to pieces.

Reply:
SystemMutex would have the additional benefit that it's a no-op when CHIP_SYSTEM_CONFIG_NO_LOCKING.... which is good as long as we actually stick to the "there is no parallel access when CHIP_SYSTEM_CONFIG_NO_LOCKING" guarantee.

Reply:
From the comment above, the primary objective of this PR seems to be moving the SystemObject code over and making it so it doesn't weirdly require pools to inherit from the Object class. Given that, it might just be better to open and issue and deal with my request here in a separate PR.

Reply:
We should actually not need locking, since the SDK is explicitly not thread-safe from outside. (Weave needed it for AsyncDNSResolver.)

We don't have any good collections. I'm giving some thought to splitting off iteration into a subclass, which would deal with the collection problem separately from the basic allocation.

Reply:
Mutex removed since the SDK is internally single-thread.